src/npfctl/npfctl.c

```diff
@@ -275,6 +278,96 @@ npfctl_print_addrmask(int alen, const char *fmt, const npf_addr_t *addr,
 	return buf;
 }
 
+static int
+npfctl_table_type(const char *typename)
+{
+	int i, type = 0;
+
+	static const struct tbltype_s {
+		const char *name;
+		u_int type;
+	} tbltypes[] = {
+		{ "ipset",	NPF_TABLE_IPSET	},
+		{ "lpm",	NPF_TABLE_LPM	},
+		{ "const",	NPF_TABLE_CONST	},
+		{ NULL,		0		}
+	};
+
+	for (i = 0; tbltypes[i].name != NULL; i++) {
+		if (strcmp(typename, tbltypes[i].name) != 0) {
+			continue;
+		}
+		type = tbltypes[i].type;
+		break;
+	}
+
+	return type;
+}
+
+__dead static void
+npfctl_table_replace(int fd, int argc, char **argv)
+{
+	const char *name, *newname, *fname;
+	nl_config_t *ncf;
+	nl_table_t *t;
+	npf_error_t errinfo;
+	int c, tid;
+	u_int type = 0;
+
+	name = newname = argv[0];
+	optind = 2;
+	while ((c = getopt(argc, argv, "n:t:")) != -1) {
+		switch (c) {
+		case 't':
+			type = npfctl_table_type(optarg);
+			if (!type) {
+				errx(EXIT_FAILURE, "unsupported table type '%s'", optarg);
+			}
+			break;
+		case 'n':
+			newname = optarg;
+			break;
+		default:
+			fprintf(stderr,
+			    "Usage: %s table <tid> replace [-n <newid>] [-t ipset|lpm|const] <table-file>\n",
+			    getprogname());
+			exit(EXIT_FAILURE);
+		}
+	}
+	argc -= optind;
+	argv += optind;
+
+	if (argc != 1) {
+		usage();
+	}
+
+	fname = argv[0];
+
+	/* Get existing config to lookup ID of existing table */
+	if (!(ncf = npf_config_retrieve(fd))) {
+		err(EXIT_FAILURE, "npf_config_retrieve()");
+	}
+	if (!(t = npfctl_table_getbyname(ncf, name))) {
+		errx(EXIT_FAILURE, "no existing table '%s' to replace", name);
```

Comment:
Actually: `"table '%s' not found in the active configuration"`

Also, can you please move this logic into a separate function, e.g. something like `npfctl_active_table_byname()`?  I will use it elsewhere..

Reply:
I had a look at doing this but it's a bit tricky to manage without introducing a memory leak; if we `npf_config_destroy()` the config, it destroys the table we're wanting to return too... 

Is there another way you can think of doing this? Two options I can think of are:
1. Add a global variable called e.g. `active_ncf`, with functions `npfctl_active_config_load()` (which wraps `npf_config_retrieve()`) and `npfctl_active_config_destroy()` to load and destroy it. Usage example:
```c
if (!npfctl_active_config_load()) {
		err(EXIT_FAILURE, "npf_active_config_load()");
}
npfctl_table_getbyname(active_ncf, tablename);
npfctl_active_config_destroy();
```

2. add a `npf_table_clone()` utility function to libnpf which wraps `nvlist_clone()`; so that the table can be cloned from the retrieved config, allowing the config to be destroyed. This would allow a `npfctl_active_table_byname()` function which does something like:
```c
nl_table_t *
npfctl_active_table_byname(const char *name)
{
	nl_config_t *ncf;
	nl_table_t *t;
	/* Get existing config to lookup ID of existing table */
	if ((ncf = npf_config_retrieve(fd)) == NULL) {
		err(EXIT_FAILURE, "npf_config_retrieve()");
	}
	if ((t = npfctl_table_getbyname(ncf, name)) != NULL) {
		t = npf_table_clone(t); /* new function which clones the nl_table_t nvlist */
	}
	npf_config_destroy(ncf);
	return t;
}
```

Neither seems an ideal solution; the first is clumsy and still requires a fair bit of inline logic; perhaps the second is a little better but it's not efficient if fetching more than one table is desired. Any preference, or other ideas?